### PR TITLE
Add Support for Blob Type (s3 and GCS)

### PR DESCRIPTION
**Pull request by khrm**
Added support for reading forwarded logs to s3 and GCS.
If vector forwards logs to s3 and GCS, these logs can be read using
Go CDK's blob library.
Also, if result logging mechanism is changed from v1alpha2 to v1alpha3
which doesn't support watcher, we can read old logs if they exist.

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you review them:

- [ ] Has [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [ ] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [ ] [Tested your changes locally](https://github.com/tektoncd/results/blob/main/docs/DEVELOPMENT.md) (if this is a code change)
- [ ] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [ ] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including functionality, content, code)
- [ ] Has a kind label. You can add a comment on this PR that contains `/kind <type>`. Valid types are bug, cleanup, design, documentation, feature, flake, misc, question, tep
- [ ] Release notes block below has been updated with any user-facing changes (API changes, bug fixes, changes requiring upgrade notices or deprecation warnings)
- [ ] Release notes contain the string "action required" if the change requires additional action from users switching to the new release

# Release Notes

```release-note
Added support for reading forwarded logs to s3 and GCS.
If vector forwards logs to s3 and GCS, these logs can be read using
Go CDK's blob library.
Also, if result logging mechanism is changed from v1alpha2 to v1alpha3
which doesn't support watcher, we can read old logs if they exist.
```

### CODE REVIEW

**Review by khrm — COMMENTED** [review]
/assign @enarha @vdemeester

File: pkg/api/server/v1alpha2/log/gcs_test.go [hunk 1]
```diff
@@ -79,6 +79,7 @@ func NewRecordReplayClient(t *testing.T, modReq func(r *httpreplay.Recorder), po
 	if err != nil {
 		t.Fatal(err)
 	}
+	rep.IgnoreHeader("X-Goog-Gcs-Idempotency-Token")
```
Comment (enarha) [inline pkg/api/server/v1alpha2/log/gcs_test.go:82]: why do we ignore this one?

File: pkg/api/server/v1alpha2/plugin/plugin_logs.go [hunk 2]
```diff
@@ -31,13 +43,32 @@ const (
...
 
+var (
+	openBucket = func(ctx context.Context, urlString string) (*blob.Bucket, error) {
+		bucket, err := blob.OpenBucket(ctx, urlString)
+		return bucket, err
+	}
+	clean = func(bucket *blob.Bucket, logger *zap.SugaredLogger) {
+		err := bucket.Close()
+		if err != nil {
+			logger.Errorf("Got error while closing blob %s", err)
```
Comment (enarha) [inline pkg/api/server/v1alpha2/plugin/plugin_logs.go:63]:
```suggestion
			logger.Errorf("Got error while closing bucket %s", err)
```

File: pkg/api/server/v1alpha2/plugin/plugin_logs.go [hunk 3]
```diff
@@ -250,8 +271,133 @@ func (s *LogPluginServer) getLokiLogs(writer io.Writer, parent string, rec *db.R
...
+	case typePipelineRun:
+		err := errors.New("pipelinerun not supported, please use taskrun")
+		s.logger.Error(err.Error())
+		return err
+	case typeTaskRun:
+		logRec := &db.Record{}
+		if legacy {
+			logRec, err = getLogRecord(s.db, parent, rec.ResultID, rec.Name)
+			if err != nil {
+				s.logger.Debugf("error getting legacy log record: %s", err)
```
Comment (enarha) [inline pkg/api/server/v1alpha2/plugin/plugin_logs.go:312]: why do we continue here?
1. If err != nil, then logRec will be nil
2. If legacy is configured which means log type to v1alpha2, why do we continue as v1alpha3?

File: pkg/api/server/v1alpha2/plugin/plugin_logs.go [hunk 4]
```diff
@@ -250,8 +271,133 @@ func (s *LogPluginServer) getLokiLogs(writer io.Writer, parent string, rec *db.R
...
+			}
+			if logRec != nil && err == nil {
+				log := &v1alpha3.Log{}
+				err := json.Unmarshal(logRec.Data, log)
+				if err != nil {
+					err = fmt.Errorf("could not decode Log record: %w", err)
+					s.logger.Error(err)
```
Comment (enarha) [inline pkg/api/server/v1alpha2/plugin/plugin_logs.go:319]: below you call the .Error() (`s.logger.Error(err.Error())`) method on the error while logging. I do not mind either, just keep consistent.
Also in the other instances below.

File: pkg/api/server/v1alpha2/plugin/server.go [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+func NewLogServer(config *config.Config, logger *zap.SugaredLogger, auth auth.Checker, db *gorm.DB) (*LogServer, error) {
+	s := &LogServer{
+		config: config,
+		logger: logger,
+		auth:   auth,
+		db:     db,
+	}
+
+	s.logger.Debugf("LOGS_TYPE: %s", strings.ToLower(s.config.LOGS_TYPE))
+	// If the logs type is not Loki, we don't need to set up the LogPluginServer
```
Comment (enarha) [inline pkg/api/server/v1alpha2/plugin/server.go:59]: setLogPlugin() with also return true if the log type BlobLogType, so either the comment is outdated or the logic below.

Comment (khrm) [reply pkg/api/server/v1alpha2/plugin/server.go:59]: Yes. Comment is outdated.

File: pkg/api/server/v1alpha2/log/gcs_test.go (hunk 1 above)
Comment (khrm) [reply pkg/api/server/v1alpha2/log/gcs_test.go:82]: This unit tests uses http replay. Earlier, when we ran this on a live cluster, GCS sdk supported this differently. Now, this header gives us error, to resolve this either we ignore this header or run against live GCS bucket and regenerate test data.

File: pkg/api/server/v1alpha2/plugin/server.go (hunk 5 above)
Comment (khrm) [reply pkg/api/server/v1alpha2/plugin/server.go]: Updated this.

File: pkg/api/server/v1alpha2/plugin/plugin_logs.go (hunk 4 above)
Comment (khrm) [reply pkg/api/server/v1alpha2/plugin/plugin_logs.go:319]: Changed it to use `s.logger.Error(err)` to everywhere.

File: pkg/api/server/v1alpha2/plugin/plugin_logs.go (hunk 3 above)
Comment (khrm) [reply pkg/api/server/v1alpha2/plugin/plugin_logs.go:312]: 1. Yes, fixed that.
2. Yes, Log Type kind itself can be v1alpha2 or  v1alpha3. 
In plugin API, there's no LogType record itself.

**Pull request merged**